File: packages/palm-engine/addressdb/addressdb.go
// Package addressdb encodes and decodes PalmOS Address Book's AddressDB.
//
// Record layout (per pilot-link libpisock/address.c):
//
//	0..3   UInt32  phoneflag = pl0<<0 | pl1<<4 | pl2<<8 | pl3<<12 |
//	                           pl4<<16 | showPhone<<20   (5 phone-slot
//	                           label indices + which phone shows in list)
//	4..7   UInt32  contents  (bit v set ⇒ entry[v] present; 19 fields)
//	8      UInt8   company string offset (for list sort; (pos-8) when
//	               the company field is written, else 0)
//	9..    present entries, index order 0..18, each NUL-terminated
//
// Field indices: 0 last, 1 first, 2 company, 3..7 phone1..5 values,
// 8 address, 9 city, 10 state, 11 zip, 12 country, 13 title,
// 14..17 custom1..4, 18 note. The phone-slot label index (0..7) selects
// a type name from phoneLabelNames.
package addressdb

import (
	"encoding/binary"

	"github.com/palmvellum/palmvellum/packages/palm-engine/charset"
	"github.com/palmvellum/palmvellum/packages/palm-engine/memodb"
	"github.com/palmvellum/palmvellum/packages/palm-engine/pdb"
)

const numEntries = 19

// Entry indices (0..18).
const (
	eLast = iota
	eFirst
	eCompany
	ePhone1 // 3..7 are the five phone-value slots
	ePhone2
	ePhone3
	ePhone4
	ePhone5
	eAddress
	eCity
	eState
	eZip
	eCountry
	eTitle
	eCustom1 // 14..17
	eCustom2
	eCustom3
	eCustom4
	eNote // 18
)

// phoneLabelNames maps a phone-slot label index (0..7) to its type name,
// matching the stock AddressDB label table.
var phoneLabelNames = [8]string{"Work", "Home", "Fax", "Other", "E-mail", "Main", "Pager", "Mobile"}

// PhoneLabelIndex returns the slot-label index for a type name, or 0 (Work).
func PhoneLabelIndex(name string) uint8 {
	for i, n := range phoneLabelNames {
		if n == name {
			return uint8(i)
		}
	}
	return 0
}

// Phone is one phone/email slot.
type Phone struct {
	Label string // Work/Home/Fax/Other/E-mail/Main/Pager/Mobile
	Value string
}

// Contact is one address-book entry.
type Contact struct {
	UniqueID  uint32
	Category  uint8
	Last      string
	First     string
	Company   string
	Title     string
	Phones    []Phone // up to 5, in slot order
	ShowPhone uint8   // which phone slot (0..4) shows in the list
	Address   string
	City      string
	State     string
	Zip       string
	Country   string
	Note      string
	Custom    [4]string
}

// DecodeContacts parses all records of an AddressDB.
func DecodeContacts(db *pdb.DB) []Contact {
	out := make([]Contact, 0, len(db.Records))
	for _, r := range db.Records {
		c, ok := decodeOne(r.Data)
		if !ok {
			continue
		}
		c.UniqueID = r.UniqueID
		c.Category = r.Attributes & 0x0F
		out = append(out, c)
	}
	return out
}

func decodeOne(b []byte) (Contact, bool) {
	if len(b) < 9 {
		return Contact{}, false
	}
	var c Contact
	phoneflag := binary.BigEndian.Uint32(b[0:4])
	pl := [5]uint8{
		uint8(phoneflag>>0) & 0x0F,
		uint8(phoneflag>>4) & 0x0F,
		uint8(phoneflag>>8) & 0x0F,
		uint8(phoneflag>>12) & 0x0F,
		uint8(phoneflag>>16) & 0x0F,
	}
	c.ShowPhone = uint8(phoneflag>>20) & 0x0F
	contents := binary.BigEndian.Uint32(b[4:8])

	entries := make([]string, numEntries)
	p := 9
	for v := 0; v < numEntries; v++ {
		if contents&(1<<uint(v)) == 0 {
			continue
		}
		raw, n := readCStr(b[p:])
		entries[v] = charset.FromPalm(raw)
		p += n
		if p > len(b) {
			p = len(b)
		}
	}

	c.Last, c.First, c.Company = entries[eLast], entries[eFirst], entries[eCompany]
	c.Address, c.City, c.State = entries[eAddress], entries[eCity], entries[eState]
	c.Zip, c.Country, c.Title = entries[eZip], entries[eCountry], entries[eTitle]
	c.Note = entries[eNote]
	c.Custom = [4]string{entries[eCustom1], entries[eCustom1+1], entries[eCustom1+2], entries[eCustom1+3]}
	for slot := 0; slot < 5; slot++ {
		val := entries[ePhone1+slot]
		if val == "" {
			continue
		}
		c.Phones = append(c.Phones, Phone{Label: phoneLabelNames[pl[slot]], Value: val})
	}
	return c, true
}

func readCStr(b []byte) (raw []byte, consumed int) {
	for i, ch := range b {
		if ch == 0 {
			return b[:i], i + 1
		}
	}
	return b, len(b)
}

// EncodeContacts turns Contacts into pdb records.
func EncodeContacts(cs []Contact) []pdb.Record {
	out := make([]pdb.Record, 0, len(cs))
	for _, c := range cs {
		out = append(out, pdb.Record{
			UniqueID:   c.UniqueID,
			Attributes: c.Category & 0x0F,
			Data:       c.encode(),
		})
	}
	return out
}

func (c Contact) encode() []byte {
	// Build the 19 entry strings (Big5).
	entries := make([][]byte, numEntries)
	set := func(i int, s string) {
		if s != "" {
			entries[i] = charset.ToPalm(s)
		}
	}
	set(eLast, c.Last)
	set(eFirst, c.First)
	set(eCompany, c.Company)
	set(eAddress, c.Address)
	set(eCity, c.City)
	set(eState, c.State)
	set(eZip, c.Zip)
	set(eCountry, c.Country)
	set(eTitle, c.Title)
	set(eNote, c.Note)
	for i := 0; i < 4; i++ {
		set(eCustom1+i, c.Custom[i])
	}

	// Phone slots + label indices.
	var pl [5]uint8
	for slot := 0; slot < 5 && slot < len(c.Phones); slot++ {
		pl[slot] = PhoneLabelIndex(c.Phones[slot].Label)
		set(ePhone1+slot, c.Phones[slot].Value)
	}

	var contents uint32
	for v := 0; v < numEntries; v++ {
		if len(entries[v]) > 0 {
			contents |= 1 << uint(v)
		}
	}

	phoneflag := uint32(pl[0]) | uint32(pl[1])<<4 | uint32(pl[2])<<8 |
		uint32(pl[3])<<12 | uint32(pl[4])<<16 | uint32(c.ShowPhone)<<20

	buf := make([]byte, 9)
	binary.BigEndian.PutUint32(buf[0:4], phoneflag)
	binary.BigEndian.PutUint32(buf[4:8], contents)

	// company offset byte (8): pos of company string within record minus 8.
	var companyOffset byte
	pos := 9
	for v := 0; v < numEntries; v++ {
		if len(entries[v]) == 0 {
			continue
		}
		if v == eCompany {
			companyOffset = byte(pos - 8)
		}
		buf = append(buf, entries[v]...)
		buf = append(buf, 0)
		pos += len(entries[v]) + 1
	}
	buf[8] = companyOffset
	return buf
}

// DisplayName is the list label the cloud stores in records.body.
func (c Contact) DisplayName() string {
	switch {
	case c.Last != "" && c.First != "":
		return c.First + " " + c.Last
	case c.Last != "":
		return c.Last
	case c.First != "":
		return c.First
	case c.Company != "":
		return c.Company
	default:
		return "(no name)"
	}
}

// NewAddressDB builds an empty AddressDB shell. appInfo (with the 22
// labels + categories) is reused verbatim from the card; pass nil for a
// bare DB (categories then come from memodb.DefaultAppInfo defaults).
func NewAddressDB(appInfo []byte) *pdb.DB {
	if appInfo == nil {
		appInfo = memodb.DefaultAppInfo().Encode()
	}
	return &pdb.DB{
		Name:    "AddressDB",
		Type:    pdb.FourCC("DATA"),
		Creator: pdb.FourCC("addr"),
		AppInfo: appInfo,
	}
}

File: packages/android-native/app/src/main/java/dev/tatliving/palmvellum/organizers/data/card/AddressDb.kt
package dev.tatliving.palmvellum.organizers.data.card

/**
 * Encode/decode PalmOS Address Book's AddressDB records — a Kotlin port of
 * `packages/palm-engine/addressdb/addressdb.go`.
 *
 * Record layout (pilot-link address.c):
 *   0-3 phoneflag = pl0..pl4 (4 bits each) | showPhone<<20
 *   4-7 contents bitmask (bit v ⇒ entry[v] present; 19 fields)
 *   8   company string offset (for list sort)
 *   9.. present entries, index order 0..18, each NUL-terminated
 * Field indices: 0 last,1 first,2 company,3..7 phone1..5,8 address,9 city,
 *   10 state,11 zip,12 country,13 title,14..17 custom1..4,18 note.
 */

private const val ADDR_NUM_ENTRIES = 19
private const val E_LAST = 0
private const val E_FIRST = 1
private const val E_COMPANY = 2
private const val E_PHONE1 = 3
private const val E_ADDRESS = 8
private const val E_CITY = 9
private const val E_STATE = 10
private const val E_ZIP = 11
private const val E_COUNTRY = 12
private const val E_TITLE = 13
private const val E_CUSTOM1 = 14
private const val E_NOTE = 18

/** AddressDB AppInfo size: 278 category block + label table + tail. */
const val ADDR_APPINFO_LEN = 638

private val PHONE_LABEL_NAMES = arrayOf("Work", "Home", "Fax", "Other", "E-mail", "Main", "Pager", "Mobile")

private fun phoneLabelIndex(name: String): Int {
    val i = PHONE_LABEL_NAMES.indexOf(name)
    return if (i >= 0) i else 0
}

class AddressPhone(var label: String = "", var value: String = "")

class Contact(
    var uniqueId: Int = 0,
    var category: Int = 0,
    var last: String = "",
    var first: String = "",
    var company: String = "",
    var title: String = "",
    var phones: MutableList<AddressPhone> = mutableListOf(), // up to 5, slot order
    var showPhone: Int = 0,
    var address: String = "",
    var city: String = "",
    var state: String = "",
    var zip: String = "",
    var country: String = "",
    var note: String = "",
    var custom: Array<String> = arrayOf("", "", "", ""),
) {
    fun displayName(): String = when {
        last.isNotEmpty() && first.isNotEmpty() -> "$first $last"
        last.isNotEmpty() -> last
        first.isNotEmpty() -> first
        company.isNotEmpty() -> company
        else -> "(no name)"
    }
}

object AddressDb {

    fun decode(db: PalmDb): List<Contact> = db.records.mapNotNull { r ->
        decodeOne(r.data)?.apply {
            uniqueId = r.uniqueId
            category = r.attributes and 0x0F
        }
    }

    private fun decodeOne(b: ByteArray): Contact? {
        if (b.size < 9) return null
        val c = Contact()
        val phoneflag = u32(b, 0)
        val pl = intArrayOf(
            ((phoneflag ushr 0) and 0x0F).toInt(),
            ((phoneflag ushr 4) and 0x0F).toInt(),
            ((phoneflag ushr 8) and 0x0F).toInt(),
            ((phoneflag ushr 12) and 0x0F).toInt(),
            ((phoneflag ushr 16) and 0x0F).toInt(),
        )
        c.showPhone = ((phoneflag ushr 20) and 0x0F).toInt()
        val contents = u32(b, 4)

        val entries = Array(ADDR_NUM_ENTRIES) { "" }
        var p = 9
        for (v in 0 until ADDR_NUM_ENTRIES) {
            if (contents and (1L shl v) == 0L) continue
            val (raw, n) = readCStr(b, p)
            entries[v] = PalmCharset.fromPalm(raw)
            p += n
            if (p > b.size) p = b.size
        }

        c.last = entries[E_LAST]; c.first = entries[E_FIRST]; c.company = entries[E_COMPANY]
        c.address = entries[E_ADDRESS]; c.city = entries[E_CITY]; c.state = entries[E_STATE]
        c.zip = entries[E_ZIP]; c.country = entries[E_COUNTRY]; c.title = entries[E_TITLE]
        c.note = entries[E_NOTE]
        c.custom = arrayOf(entries[E_CUSTOM1], entries[E_CUSTOM1 + 1], entries[E_CUSTOM1 + 2], entries[E_CUSTOM1 + 3])
        for (slot in 0 until 5) {
            val value = entries[E_PHONE1 + slot]
            if (value.isEmpty()) continue
            c.phones.add(AddressPhone(PHONE_LABEL_NAMES[pl[slot]], value))
        }
        return c
    }

    fun encode(contacts: List<Contact>): MutableList<PalmRecord> = contacts.map { c ->
        PalmRecord(uniqueId = c.uniqueId, attributes = c.category and 0x0F, data = encodeOne(c))
    }.toMutableList()

    private fun encodeOne(c: Contact): ByteArray {
        val entries = arrayOfNulls<ByteArray>(ADDR_NUM_ENTRIES)
        fun set(i: Int, s: String) { if (s.isNotEmpty()) entries[i] = PalmCharset.toPalm(s) }
        set(E_LAST, c.last); set(E_FIRST, c.first); set(E_COMPANY, c.company)
        set(E_ADDRESS, c.address); set(E_CITY, c.city); set(E_STATE, c.state)
        set(E_ZIP, c.zip); set(E_COUNTRY, c.country); set(E_TITLE, c.title); set(E_NOTE, c.note)
        for (i in 0 until 4) set(E_CUSTOM1 + i, c.custom.getOrElse(i) { "" })

        val pl = IntArray(5)
        var slot = 0
        while (slot < 5 && slot < c.phones.size) {
            pl[slot] = phoneLabelIndex(c.phones[slot].label)
            set(E_PHONE1 + slot, c.phones[slot].value)
            slot++
        }

        var contents = 0L
        for (v in 0 until ADDR_NUM_ENTRIES) if (entries[v] != null) contents = contents or (1L shl v)

        val phoneflag = (pl[0].toLong()) or (pl[1].toLong() shl 4) or (pl[2].toLong() shl 8) or
            (pl[3].toLong() shl 12) or (pl[4].toLong() shl 16) or (c.showPhone.toLong() shl 20)

        val head = ByteArray(9)
        putU32(head, 0, phoneflag)
        putU32(head, 4, contents)
        val body = java.io.ByteArrayOutputStream()
        var companyOffset = 0
        var pos = 9
        for (v in 0 until ADDR_NUM_ENTRIES) {
            val e = entries[v] ?: continue
            if (v == E_COMPANY) companyOffset = pos - 8
            body.write(e); body.write(0)
            pos += e.size + 1
        }
        head[8] = companyOffset.toByte()
        return head + body.toByteArray()
    }

    /** Stock 638-byte AppInfo (categories + 22 field labels); used only when the card has none. */
    fun defaultAppInfo(): ByteArray {
        val buf = ByteArray(ADDR_APPINFO_LEN)
        val mem = MemoAppInfo.default()
        mem.categories[1] = MemoCategory("Business", 1, true)
        mem.categories[2] = MemoCategory("Personal", 2, true)
        mem.categories[3] = MemoCategory("QuickList", 3, true)
        mem.encode().copyInto(buf, 0, 0, 278)
        for (i in STD_LABELS.indices) {
            val off = 282 + i * 16
            val l = STD_LABELS[i].toByteArray(Charsets.US_ASCII)
            l.copyInto(buf, off, 0, minOf(l.size, 15))
        }
        return buf
    }

    /**
     * Empty AddressDB shell. The card's own AppInfo (638 bytes, with the 22
     * field labels) is reused verbatim; anything shorter is rebuilt from stock
     * so the Palm restore won't fault (a 280-byte memo-shaped block crashes it).
     */
    fun newDb(appInfo: ByteArray): PalmDb {
        val ai = if (appInfo.size < ADDR_APPINFO_LEN) defaultAppInfo() else appInfo
        return PalmDb(
            name = "AddressDB",
            type = "DATA".toByteArray(Charsets.US_ASCII),
            creator = "addr".toByteArray(Charsets.US_ASCII),
            appInfo = ai,
        )
    }

    private val STD_LABELS = arrayOf(
        "Last name", "First name", "Company", "Work", "Home", "Fax", "Other",
        "E-mail", "Address", "City", "State", "Zip Code", "Country", "Title",
        "Custom 1", "Custom 2", "Custom 3", "Custom 4", "Note", "Main", "Pager", "Mobile",
    )

    private fun u32(b: ByteArray, o: Int): Long =
        ((b[o].toLong() and 0xFF) shl 24) or ((b[o + 1].toLong() and 0xFF) shl 16) or
            ((b[o + 2].toLong() and 0xFF) shl 8) or (b[o + 3].toLong() and 0xFF)

    private fun putU32(b: ByteArray, o: Int, v: Long) {
        b[o] = (v ushr 24).toByte(); b[o + 1] = (v ushr 16).toByte()
        b[o + 2] = (v ushr 8).toByte(); b[o + 3] = v.toByte()
    }
}
